### PR TITLE
migrate to KMP

### LINKED ISSUE #88: Support Compose Multiplatform
I would like to use this library in a multi-platform project. As the library does not seem to have any dependency on Android it should be as easy as restructuring and changing the config. I'm happy to contribute the restructuring part. Publishing may require some changes, which will need to be taken care of.

And thanks for this amazing project :)

### PR DESCRIPTION
**Pull request by ShreyashKore**
closes #88

### CODE REVIEW

**Comment by ShreyashKore** [comment]
I haven't set up the publishing part; as I lack knowledge of the maven-central publishing. I've also kept both nexus and vanniktech publishing plugins in `lib.versions.toml`. We can remove one of them once you decide which one to use.

File: sample/src/commonMain/composeResources/drawable/ic_cyclone.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (snappdevelopment) [inline sample/src/commonMain/composeResources/drawable/ic_cyclone.xml:1]: The following four icons seem to be unused. What are they used for?

File: sample/src/commonMain/composeResources/values/strings.xml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<resources>
```
Comment (snappdevelopment) [inline sample/src/commonMain/composeResources/values/strings.xml:1]: This is empty and could be removed

**Review by snappdevelopment — CHANGES_REQUESTED** [review]
Sorry, it took me a while to review your changes. I checked out your branch and tried to run everything. I came across a couple of issues:

1. I tried to run the desktop app from the gutter icon, which doesn't seem to work. Maybe it's a tooling issue, but it also throws an exception that it can't find `androidx/collection/ScatterMapKt`. However running the app with `./gradlew run` works.
2. When I try to scroll the json in the desktop app it doesn't work and prints an IndexOutOfBoundsException.
3. I am not sure how to run the iOS app. I will have to look into that. Were you able to run it on your machine?
4. I can't run the UI tests. If I do `./gradlew :composeApp:connectedAndroidTest` like the docs say, it tells me: `Cannot locate tasks that match ':composeApp:connectedAndroidTest' as project 'composeApp' not found in root project 'JsonTree'.`. Do we need to move the tests to the sample package where the actual app is?

The sample app also has a couple of styling issues, but I can fix that later

File: jsontree/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+kotlin {
+    jvm()
+    androidTarget {
+        publishLibraryVariants("release")
+        @OptIn(ExperimentalKotlinGradlePluginApi::class)
+        instrumentedTestVariant.sourceSetTree.set(KotlinSourceSetTree.test)
+        @OptIn(ExperimentalKotlinGradlePluginApi::class)
+        compilerOptions {
+            jvmTarget.set(JvmTarget.JVM_1_8)
```
Comment (ShreyashKore) [inline jsontree/build.gradle.kts:22]: We may need to remove either `androidTarget. compilerOptions` or `jvmToolchain` block below.

**Comment by ShreyashKore** [comment]
I'm sorry for missing your questions earlier. Went straight to the requested changes 😅 

> 1. I tried to run the desktop app from the gutter icon, which doesn't seem to work. Maybe it's a tooling issue, but it also throws an exception that it can't find `androidx/collection/ScatterMapKt`. However running the app with `./gradlew run` works.

For me, running from the gutter icon is working. This is the command being used when running from the gutter icon `./gradlew sample:jvmRun -DmainClass=MainKt --quiet`. I'm using OpenJDK version "17.0.10"

> 2. When I try to scroll the json in the desktop app it doesn't work and prints an IndexOutOfBoundsException.

I'm getting no such exception on the Desktop or any other platform.

> 3. I am not sure how to run the iOS app. I will have to look into that. Were you able to run it on your machine?

I've added `iosApp` module. I'm running the project on mac os, so I am able to build the ios app.

> 4. I can't run the UI tests. If I do `./gradlew :composeApp:connectedAndroidTest` like the docs say, it tells me: `Cannot locate tasks that match ':composeApp:connectedAndroidTest' as project 'composeApp' not found in root project 'JsonTree'.`. Do we need to move the tests to the sample package where the actual app is?

Running Android Tests from `commonTest` doesn't seem to be working for now in CMP. You can still run the JVM tests (It seems tests are more flaky on JVM) currently one is failing 😞).
Use gutter icon to run individual test or use `./gradlew :jsontree:jvmTest` command (note the module name is not `composeApp` but `jsontree`).

> The sample app also has a couple of styling issues, but I can fix that later

Let me know how I can help.

**Comment by snappdevelopment** [comment]
> For me, running from the gutter icon is working. This is the command being used when running from the gutter icon ./gradlew sample:jvmRun -DmainClass=MainKt --quiet. I'm using OpenJDK version "17.0.10"

It looks like Zac Sweers got the same exception in a Compose Multiplatform project, but it seems like there is no solution. Since running the command works, that's fine for me.

> I've added iosApp module. I'm running the project on mac os, so I am able to build the ios app.

Thanks for adding it. I also tried that but I couldn't configure my run config correctly, because my MacOS and Xcode are too old. I tried it on a different Mac and it worked.

> Running Android Tests from commonTest doesn't seem to be working for now in CMP. You can still run the JVM tests including UI tests (It seems tests are more flaky on JVM currently one is failing 😞).
Use gutter icon to run individual test or use ./gradlew :jsontree:jvmTest command (note the module name is not composeApp but jsontree).

That's unfortunate, but okay. `./gradlew :jsontree:jvmTest` runs successful for me.

One more question:
Do you also get those warnings? Since the iOS app works, I assume this is a false positive?
```
w: A compileOnly dependency is used in the Kotlin/Native target 'iosArm64':
Compilation: main
Dependencies:
org.jetbrains.kotlinx:atomicfu:0.22.0
w: A compileOnly dependency is used in the Kotlin/Native target 'iosSimulatorArm64':
Compilation: main
Dependencies:
org.jetbrains.kotlinx:atomicfu:0.22.0
w: A compileOnly dependency is used in the Kotlin/Native target 'iosX64':
Compilation: main
Dependencies:
org.jetbrains.kotlinx:atomicfu:0.22.0

```

File: jsontree/build.gradle.kts (hunk 3 above)
Comment (snappdevelopment) [reply jsontree/build.gradle.kts:22]: Let's remove `androidTarget. compilerOptions`

**Comment by ShreyashKore** [comment]
Yes, I'm getting the same "compileOnly dependency" warnings.. but according to this https://github.com/Kotlin/kotlinx-atomicfu/issues/448 issue, they are harmless.

**Comment by snappdevelopment** [comment]
I will go ahead and merge this now. I will do some more adjustments to the sample and CI and then try to publish the new version. Thanks for contributing 👍

**Comment by snappdevelopment** [comment]
@ShreyashKore I just released a new snapshot version. Can you confirm if it works for you?

The version is `2.3.0-SNAPSHOT`
The snapshot repository can be added like this in your `dependencyResolutionManagement` block in `settings.gradle.kts`:
```
maven { url = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/") }
```

**Comment by ShreyashKore** [comment]
It's working on jvmDesktop (mac os)  and Android. On IOS, it's giving me some errors... Not sure if it's something with the library or the fact that I'm using an experimental version of Compose... I will try to pinpoint the issue and will attach the stack traces.

**Comment by snappdevelopment** [comment]
Yes, it seems like it's a compatibility issue with the pre-release version of Compose. The iOS app worked for me, so I googled the error and saw that other people had similar problems. They were able to fix it by using a compatible Compose version.
I will try to reproduce this, but I would prefer to stay on a stable Compose version for the release.

**Comment by ShreyashKore** [comment]
Yup... Just checked on a fresh project with Kotlin "2.0.20" and compose 1.6.11 and it is working fine 👍.

**Pull request merged**